### DISCUSSION TITLE
Error in Rust — learning Rust

### DISCUSSION THREAD

**Discussion by ghost** [community/community#142285] [Programming Help]
### Body

```rust
fn main(){
    let list = &["apple", "oranges", "banana"];
    let fruit: str = "apple";
    for fruit in list{
        println!("I like {} ", list[fruit])
    }
}
```

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by YuheshPandian** [comment]


**Reply by YuheshPandian** [reply]
I'm a beginner in terms of rust i hope this works

**Comment by spenserblack** [comment]
tl;dr here's the fixed code

```rust
fn main() {
    let list = &["apple", "oranges", "banana"];
    let fruit: &str = "apple";
    for fruit in list{
        println!("I like {} ", fruit);
    }
}
```

It's generally helpful to include the error messages when you're looking for help. And the Rust compiler does a great job of providing helpful error messages.

> ```rust
> let fruit: str = "apple";
> ```

This results in the following error message:

```
3 |     let fruit: str = "apple";
  |                ---   ^^^^^^^ expected `str`, found `&str`
  |                |
  |                expected due to this
error[E0277]: the size for values of type `str` cannot be known at compilation time
...
```

You might be wondering what Rust means when it complains about `str` not having a known size It's a pretty strange concept if you come from using other languages. But, regardless, the error message provides exactly what you need to do -- replace `str` with `&str`.

The reason that `str` doesn't have a known size is that it can be `""`, `"foo"`, or *anything.* But `&str` is a *reference* to a `str`, and references have known sizes.

> ```rust
> for fruit in list{
>     println!("I like {} ", list[fruit])
> }
> ```

TBH I'm not really sure why it was written like this -- were you expecting `fruit` to be an index here? `for VAL in ITERABLE` will iterate over the *items,* not the indices. This is true of other languages, too, like Python and Ruby. Perhaps you were expecting that due to JavaScript's `for (const VAL in ITERABLE)` behavior?

Hope this helps! :slightly_smiling_face:
Pay close attention to Rust's error messages, they'll often tell you exactly what you need to know.

**Reply by ghost** [reply]
I figured it out it is:
```rust
let fruits = &[“apples”, “bananas”]
let fruit: str = “apples”;
if fruits.contains(fruit) {
  println!(fruit);
}
```
